nixos/modules/services/networking/envoy.nix

```diff
@@ -0,0 +1,76 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.envoy;
+
+in
+
+{
+  options.services.envoy = {
+    enable = mkEnableOption "Envoy reverse proxy";
+
+    settings = mkOption {
+      type = (pkgs.formats.json { }).type;
+      default = { };
+      example = literalExpression ''
+        {
+          admin = {
+            access_log_path = "/dev/null";
+            address = {
+              socket_address = {
+                protocol = "TCP";
+                address = "127.0.0.1";
+                port_value = 9901;
+              };
+            };
+          };
+          static_resources = {
+            listeners = [];
+            clusters = [];
+          };
+        }
+      '';
+      description = ''
+        Specify the configuration for Envoy in Nix.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.envoy ];
+    users.groups._envoy = { };
+    users.users._envoy = {
+      description = "Envoy service user";
+      group = "_envoy";
+      isSystemUser = true;
+    };
```

Comment:
can this service run with `DynamicUser`? if so we could just skip allocating global uids/gids.

Reply:
with my current configuration, yes. But I am not sure how it would be handled if a user wanted to maintain access/error logs in files (rather than being sent to stdout/err as in my current config). I was concerned that might lead to issues with dynamic user/group. Also can the cache be reused across reboots with a dynamic user/group?

Reply:
depends on how likely it is that people will use a different log dir. if it's likely the module should handle it transparently (eg bind-mounting the provided directory to `/var/log/envoy`), if it isn't likely then it's easy enough for users to add a bind mount to the service definition from their own configurations :)

Reply:
Alright I figure most people will use journal, some will use /var/log/envoy, and a small number will use something different. They can probably figure it out for themselves. Added DynamicUser = true and LogsDirectory = "envoy".